File: apps/website/src/server/db/seeds/dividir-estudiantes-en-grupos/solution-dynamicProgramming.cpp
#include <vector>

using namespace std;

long aux(vector<vector<long>> &memo, long cantEstudiantes, long cantGrupos) {
  if (cantEstudiantes == cantGrupos || cantGrupos == 1) {
    return 1;
  }
  if (memo[cantEstudiantes][cantGrupos] == -1) {
    memo[cantEstudiantes][cantGrupos] =
        cantGrupos * aux(memo, cantEstudiantes - 1, cantGrupos) +
        aux(memo, cantEstudiantes - 1, cantGrupos - 1);
  }
  return memo[cantEstudiantes][cantGrupos];
};

long dividirEstudiantesEnGrupos(long cantEstudiantes, long cantGrupos) {
  long rows = cantEstudiantes + 1;
  long columns = cantGrupos + 1;
  vector<vector<long>> memo(rows, vector<long>(columns, -1));

  return aux(memo, cantEstudiantes, cantGrupos);
}

File: apps/website/src/server/db/seeds/dividir-estudiantes-en-grupos/solution-dynamicProgramming.py
#
# Completar la función 'dividirEstudiantesEnGrupos' de acá abajo.
#
# Esta función retorna un número entero.
#
def aux(memo, cantEstudiantes, cantGrupos):
    if cantEstudiantes == cantGrupos or cantGrupos == 1:
        return 1
    if memo[cantEstudiantes][cantGrupos] is None:
        memo[cantEstudiantes][cantGrupos] = cantGrupos * aux(memo, cantEstudiantes - 1, cantGrupos) + aux(memo, cantEstudiantes - 1, cantGrupos - 1)
    return memo[cantEstudiantes][cantGrupos]

def dividirEstudiantesEnGrupos(cantEstudiantes, cantGrupos):
    rows = cantEstudiantes + 1
    columns = cantGrupos + 1
    memo = [[None] * columns for _ in range(rows)]
    return aux(memo, cantEstudiantes, cantGrupos)
